django/db/backends/base/base.py

```diff
@@ -641,3 +648,33 @@ def copy(self, alias=None, allow_thread_sharing=None):
         if allow_thread_sharing is None:
             allow_thread_sharing = self.allow_thread_sharing
         return type(self)(settings_dict, alias, allow_thread_sharing)
+
+    def _push_execute_hook(self, hook, for_many=None):
+        self.execute_hooks.append((hook, for_many))
+
+    def _pop_execute_hook(self):
+        return self.execute_hooks.pop()
+
+    @contextmanager
+    def execute_hook(self, hook, for_many=None):
+        """
+        Return a context manager, under which the hook is applied to suitable
+        database query executions
+        """
+        self._push_execute_hook(hook, for_many)
+        yield
+        self._pop_execute_hook()
```

Comment:
Should probably do a `try...finally` to ensure the hook is removed.

Reply:
Yes, it should. I ran into that while testing the new implementation.